### PR TITLE
feat(client): Stringify WebChannel payloads in Fx Desktop >= 50.

### LINKED ISSUE #4577: FxAccounts should send strings through webchannels
Originally filed as https://bugzilla.mozilla.org/show_bug.cgi?id=1275616

Firefox >= 50 expects all WebChannel payloads to be strings, not objects.

### PR DESCRIPTION
**Pull request by shane-tomlinson**
Fx Desktop >= 50 expects to see the `detail` property of WebChannel
message payloads sent as strings. This makes it so.

fixes #4577 

@rfk, @mhammond - r?

@mhammond - I was unsure whether Fennec >= 50 has the same behavior, so I only send strings for Fx Desktop at the moment.

### CODE REVIEW

File: app/scripts/lib/channels/senders/web-channel.js [hunk 1]
```diff
@@ -11,35 +11,47 @@ define(function (require, exports, module) {
...
+    initialize (options = {}) {
       this._window = options.window;
       this._webChannelId = options.webChannelId;
     },
 
+    /**
+     * Send a WebChannel message.
+     *
+     * @param {String} command command name
+     * @param {Objet} data payload
```
Comment (vladikoff) [inline app/scripts/lib/channels/senders/web-channel.js]: `Objet` -> `Object`

File: app/scripts/lib/channels/senders/web-channel.js [hunk 2]
```diff
@@ -54,20 +66,57 @@ define(function (require, exports, module) {
...
+  /**
+   * Format the `detail` payload for the current browser.
+   *
+   * @param {Object} win receiving window
+   * @param {Object} eventDetail `detail` payload
+   * @returns {String|Object}
+   */
+  function formatEventDetail(win, eventDetail) {
+    const userAgent = new UserAgent(win.navigator.userAgent);
+    // Firefox Desktop >= 50 expects the detail to be sent as a string.
```
Comment (vladikoff) [inline app/scripts/lib/channels/senders/web-channel.js]: is it only FF Desktop or does this also apply to Fennec (Android)?

File: app/scripts/lib/channels/senders/web-channel.js [hunk 3]
```diff
@@ -54,20 +66,57 @@ define(function (require, exports, module) {
...
   }
 
-  module.exports = WebChannelSender;
-});
+  /**
+   * Create an object for the `detail` field of a WebChannelMessageToChrome event.
+   *
+   * @param {String} webChannelId ID of the receiving WebChannel
+   * @param {String} command command name
+   * @param {Objet} data payload
```
Comment (vladikoff) [inline app/scripts/lib/channels/senders/web-channel.js]: Objet - > Object

File: app/scripts/lib/channels/senders/web-channel.js (hunk 2 above)
Comment (shane-tomlinson) [inline app/scripts/lib/channels/senders/web-channel.js]: That's what I asked @mhammond

File: app/tests/spec/lib/channels/senders/web-channel.js [hunk 4]
```diff
@@ -49,6 +78,23 @@ define(function (require, exports, module) {
             assert.equal(eventData.message.data.key, 'value');
           });
       });
+    }
+
+    describe('send', function () {
+      describe('Fx Dekstop >= 50', () => {
```
Comment (vladikoff) [inline app/tests/spec/lib/channels/senders/web-channel.js]: `Dekstop` -> `Desktop`

File: app/scripts/lib/channels/senders/web-channel.js (hunk 2 above)
Comment (mhammond) [inline app/scripts/lib/channels/senders/web-channel.js]: IIUC, Fennec uses the same WebChannel.jsm as desktop. What I don't know is what Fennec version it landed in :( I asked on #mobile and I'll let you know what response I get.

(and @thomcc ended up answering - it is 50 on Fennec too - see bug 1287884)

File: app/scripts/lib/channels/senders/web-channel.js [hunk 5]
```diff
@@ -54,20 +66,57 @@ define(function (require, exports, module) {
...
+   * @param {Object} win receiving window
+   * @param {Object} eventDetail `detail` payload
+   * @returns {String|Object}
+   */
+  function formatEventDetail(win, eventDetail) {
+    const userAgent = new UserAgent(win.navigator.userAgent);
+    // Firefox Desktop >= 50 expects the detail to be sent as a string.
+    // See https://bugzilla.mozilla.org/show_bug.cgi?id=1275616 and
+    // https://bugzilla.mozilla.org/show_bug.cgi?id=1238128
+    if (userAgent.isFirefoxDesktop() && userAgent.browser.version >= 50) {
```
Comment (mhammond) [inline app/scripts/lib/channels/senders/web-channel.js]: As we mentioned in bugzilla, another alternative would be to just wait a few months and lose the UA check completely :)

**Pull request merged**